tests/lean/run/structuralMutual.lean

```diff
@@ -0,0 +1,574 @@
+
+theorem eq_true_of_not_eq_false {b : Bool} : (! b) = false → b = true := by simp
+theorem eq_false_of_not_eq_true {b : Bool} : (! b) = true → b = false := by simp
+
+/--
+info: n : Nat
+h : EvenOdd.isOdd (n + 1) = false
+⊢ EvenOdd.isEven n = true
+-/
+#guard_msgs in
+theorem ex1 (n : Nat) (h : isEven (n+2) = true) : isEven n = true := by
+  replace h := eq_false_of_not_eq_true h
+  trace_state -- without smart unfolding the state would be a mess
+  replace h := eq_true_of_not_eq_false h
+  exact h
+
+end EvenOdd
+
+namespace MutualIndNonMutualFun
+
+mutual
+inductive A
+  | self : A → A
+  | other : B → A
+  | empty
+inductive B
+  | self : B → B
+  | other : A → B
+  | empty
+end
+
+-- Structural recursion ignoring some types of the mutual inductive
+
+def A.self_size : A → Nat
+  | .self a => a.self_size + 1
+  | .other _ => 0
+  | .empty => 0
+termination_by structural x => x
+
+#guard_msgs in
+def B.self_size : B → Nat
+  | .self b => b.self_size + 1
+  | .other _ => 0
+  | .empty => 0
+termination_by structural x => x
+
+-- Structural recursion with more than one function per types of the mutual inductive
+
+mutual
+def A.weird_size1 : A → Nat
+  | .self a => a.weird_size2 + 1
+  | .other _ => 0
+  | .empty => 0
+termination_by structural x => x
+def A.weird_size2 : A → Nat
+  | .self a => a.weird_size3 + 1
+  | .other _ => 0
+  | .empty => 0
+termination_by structural x => x
+def A.weird_size3 : A → Nat
+  | .self a => a.weird_size1 + 1
+  | .other _ => 0
+  | .empty => 0
+termination_by structural x => x
+end
+
+-- We have equality
+theorem A.weird_size1_eq_1 (a : A) : (A.self a).weird_size1 = a.weird_size2 + 1 := rfl
+
+-- And the right equational theorems
+/--
+info: MutualIndNonMutualFun.A.weird_size1.eq_1 (a : A) : a.self.weird_size1 = a.weird_size2 + 1
+-/
+#guard_msgs in
+#check A.weird_size1.eq_1
+
+end MutualIndNonMutualFun
+
+namespace NestedWithTuple
+
+inductive Tree where
+  | leaf
+  | node : (Tree × Tree) → Tree
+
+def Tree.below_1 (motive : Tree → Sort u) : Tree → Sort (max 1 u) :=
+  @Tree.below motive (fun _tt => PUnit)
+
+-- Assume we had this construction:
+@[reducible] protected noncomputable def Tree.brecOn.{u}
+  {motive : Tree → Sort u}
+  (t : Tree)
+  (F : (t : Tree) → Tree.below_1 motive t → motive t) :
+  motive t :=
+  let motive_below := fun t => PProd (motive t) (Tree.below_1 motive t)
+  (@Tree.rec
+    motive_below
+    -- This is the hypthetical `Pair_Tree.below tt` unfolded
+    (fun ⟨t₁,t₂⟩ => PProd PUnit.{u} (PProd (motive_below t₁) (PProd (motive_below t₂) PUnit)))
+    ⟨F Tree.leaf PUnit.unit, PUnit.unit⟩
+    (fun ⟨a₁,a₂⟩ a_ih => ⟨F (Tree.node ⟨a₁, a₂⟩) ⟨a_ih, PUnit.unit⟩, ⟨a_ih, PUnit.unit⟩⟩)
+    (fun _a _a_1 a_ih a_ih_1 => ⟨PUnit.unit, ⟨a_ih, ⟨a_ih_1, PUnit.unit⟩⟩⟩)
+    t).1
+
+-- Then the decrecursifier works just fine! (and FunInd too, see below)
+#guard_msgs in
+def Tree.size : Tree → Nat
+  | leaf => 0
+  | node (t₁, t₂) => t₁.size + t₂.size
+termination_by structural t => t
```

Comment:
@nomeata, we started using this patch and it has already made some proofs easier. Thank you!

I'm curious, do you plan to add support for nested mutuals as well? We have both nested mutuals as well as mutual functions that also use inductive that are not part of a mutual block and are interested in beta-test some of your features. Interesting to use are the following mutuals:


# Only Expr and Com are mutual inductive types, but HVector is also part of the mutual functions.
https://github.com/opencompl/lean-mlir/blob/97f33691aec49b7b939ec0452b835e93c0a18adc/SSA/Core/Framework.lean#L523

# Termination due to well-foundedness requires high heartbeats, but our mutual recursion is a bit complex. We would really like to use mutual recursion here, if possible.

https://github.com/opencompl/lean-mlir/blob/97f33691aec49b7b939ec0452b835e93c0a18adc/SSA/Core/Framework.lean#L1720

Reply:
I plan to add support for (possibly mutual) structural recursion over nested inductive data types, but not nested recursion.

In other words of you have `inductive Tree where | node : List Tree -> Tree` you will be able to define recursive functions over it, but you cannot use `List.map` but have to replicate that functionality in the mutual group.

Reply:
Lovely. Thank you for the update.